File: tests/api_recovery/RecoveryProbe.kt
import uniffi.cera_ffi.BackendPreference
import uniffi.cera_ffi.EngineConfig
import uniffi.cera_ffi.FfiConverterTypeSessionRecoveryStatus
import uniffi.cera_ffi.FfiException
import uniffi.cera_ffi.FinishReason
import uniffi.cera_ffi.GenerateOpts
import uniffi.cera_ffi.IngestRecovery
import uniffi.cera_ffi.KvCompression
import uniffi.cera_ffi.KvRewindFailure
import uniffi.cera_ffi.ModalitySink
import uniffi.cera_ffi.ModelLoader
import uniffi.cera_ffi.ModelSource
import uniffi.cera_ffi.RecoveryOutcome
import uniffi.cera_ffi.Session
import uniffi.cera_ffi.SessionConfig
import uniffi.cera_ffi.SessionRecoveryStatus
import uniffi.cera_ffi.UserMessage

private class ProbeSink(
    val session: Session,
) : ModalitySink {
    var busy = 0
    var unexpected = 0
    val done = mutableListOf<FinishReason>()

    override fun onThoughtChunk(text: String) {}

    override fun onAudioFrames(
        pcm: List<Float>,
        sampleRate: UInt,
    ) {}

    override fun onTextChunk(text: String) {
        try {
            session.recoveryStatus()
            unexpected++
        } catch (_: FfiException.Busy) {
            busy++
        } catch (_: Exception) {
            unexpected++
        }
    }

    override fun onDone(reason: FinishReason) {
        done.add(reason)
    }
}

fun main(args: Array<String>) {
    require(args.size == 2)
    val backend =
        when (args[1]) {
            "cpu" -> BackendPreference.CPU
            "metal" -> BackendPreference.METAL
            "wgpu" -> BackendPreference.GPU
            else -> error("unknown backend")
        }
    val wide = (1uL shl 40) + 7uL
    val reasons =
        listOf(
            KvRewindFailure.OutOfBounds(wide, 3uL),
            KvRewindFailure.Compressed,
            KvRewindFailure.NonCausal,
            KvRewindFailure.MissingConvolutionCheckpoint(2uL, wide),
            KvRewindFailure.InvalidCacheLayout(5uL, "unequal rows"),
            KvRewindFailure.BackendUnsupported,
            KvRewindFailure.Unknown("future rewind cause"),
        )
    for (reason in reasons) {
        val original =
            SessionRecoveryStatus(
                false,
                3u,
                IngestRecovery(
                    RecoveryOutcome.UNUSABLE,
                    reason,
                    FfiException.OutOfMemory(wide),
                ),
            )
        val decoded = FfiConverterTypeSessionRecoveryStatus.lift(FfiConverterTypeSessionRecoveryStatus.lower(original))
        check(!decoded.usable && decoded.position == 3u)
        check(decoded.lastIngestRecovery?.outcome == RecoveryOutcome.UNUSABLE)
        check(decoded.lastIngestRecovery?.rewindError == reason)
        check((decoded.lastIngestRecovery?.resetError as FfiException.OutOfMemory).requestedBytes == wide)
    }
    var cases = 0
    val compressed = KvCompression.TurboQuant(7uL, true, true)
    for (compression in listOf(KvCompression.None, KvCompression.F16, compressed)) {
        for (operation in 0..2) {
            ModelLoader(ModelSource.Path(args[0]), EngineConfig(backend = backend)).use { loader ->
                loader.buildGenerative().use { model ->
                    ModelLoader(ModelSource.Path(args[0]), EngineConfig(backend = backend)).use { referenceLoader ->
                        referenceLoader.buildGenerative().use { referenceModel ->
                            val config = SessionConfig(maxSeqLen = 32u, seed = 1361uL, ubatchSize = 1u, kvCompression = compression)
                            model.createSession(config).use { actual ->
                                referenceModel.createSession(config).use { reference ->
                                    check(actual.recoveryStatus().lastIngestRecovery == null)
                                    actual.appendTokens(listOf(0u, 1u))
                                    try {
                                        actual.sendMessage(UserMessage(text = "a".repeat(64)))
                                        error("expected overflow")
                                    } catch (
                                        e: FfiException.ContextOverflow,
                                    ) {
                                        check(e.maxSeqLen == 32u && e.by == 34u)
                                    }
                                    val unchanged = actual.recoveryStatus()
                                    check(
                                        unchanged.usable && unchanged.position == 2u &&
                                            unchanged.lastIngestRecovery?.outcome == RecoveryOutcome.UNCHANGED,
                                    )
                                    actual.appendTokens(listOf(1u))
                                    check(actual.recoveryStatus().lastIngestRecovery?.outcome == RecoveryOutcome.UNCHANGED)
                                    actual.reset()
                                    check(actual.recoveryStatus().lastIngestRecovery == null)
                                    actual.appendTokens(listOf(0u, 1u))
                                    actual.cancel()
                                    val message = UserMessage(text = "baba")
                                    val failedSink = ProbeSink(actual)
                                    try {
                                        when (operation) {
                                            0 -> actual.sendMessage(message)
                                            1 -> actual.sendMessageAndGenerate(message, GenerateOpts(maxTokens = 4u))
                                            else -> actual.sendMessageStreaming(message, GenerateOpts(maxTokens = 4u), failedSink)
                                        }
                                        error("expected cancellation")
                                    } catch (_: FfiException.Cancelled) {
                                    }
                                    val status = actual.recoveryStatus()
                                    val expected =
                                        if (args[1] == "cpu" &&
                                            compression != compressed
                                        ) {
                                            RecoveryOutcome.RESTORED
                                        } else {
                                            RecoveryOutcome.RESET
                                        }
                                    check(status.usable && status.lastIngestRecovery?.outcome == expected)
                                    check(status.position == if (expected == RecoveryOutcome.RESTORED) 2u else 0u)
                                    check(status.lastIngestRecovery?.resetError == null)
                                    if (expected == RecoveryOutcome.RESET) {
                                        check(
                                            status.lastIngestRecovery?.rewindError ==
                                                if (args[1] == "cpu") KvRewindFailure.Compressed else KvRewindFailure.BackendUnsupported,
                                        )
                                    }
                                    if (operation == 2) check(failedSink.done == listOf(FinishReason.Cancelled))
                                    try {
                                        actual.sendMessage(message)
                                        error("cancel latch cleared")
                                    } catch (_: FfiException.Cancelled) {
                                    }
                                    actual.clearCancel()
                                    if (expected == RecoveryOutcome.RESET) actual.appendTokens(listOf(0u, 1u))
                                    reference.appendTokens(listOf(0u, 1u))
                                    for (session in listOf(actual, reference)) session.sendMessage(message)
                                    check(actual.recoveryStatus().lastIngestRecovery == null)
                                    check(status.lastIngestRecovery?.outcome == expected)
                                    val options = GenerateOpts(maxTokens = 4u, temperature = 1.0f)
                                    val actualTokens = actual.generate(options).tokens
                                    val referenceTokens = reference.generate(options).tokens
                                    check(actualTokens == referenceTokens && actualTokens.size == 4)
                                    // A same-length wrong prefix must be observable, or token parity is vacuous.
                                    reference.reset()
                                    reference.appendTokens(listOf(1u, 0u))
                                    reference.sendMessage(message)
                                    val wrongPrefixTokens = reference.generate(options).tokens
                                    check(actualTokens != wrongPrefixTokens) { "continuation oracle cannot detect a wrong prefix" }
                                    val sink = ProbeSink(actual)
                                    actual.generateStreaming(GenerateOpts(maxTokens = 2u), sink)
                                    check(sink.busy > 0 && sink.unexpected == 0 && sink.done.size == 1)
                                    cases++
                                }
                            }
                        }
                    }
                }
            }
        }
    }
    println("passed $cases recovery cases: ${args[1]}")
}

File: tests/api_recovery/RecoveryProbe.swift
import Cera
import Foundation

private func require(_ condition: Bool, _ detail: String) throws {
  if !condition {
    throw NSError(domain: "RecoveryProbe", code: 1, userInfo: [NSLocalizedDescriptionKey: detail])
  }
}

private final class ProbeSink: ModalitySink, @unchecked Sendable {
  let session: Session
  var busy = 0
  var unexpected = 0
  var done: [FinishReason] = []
  init(_ session: Session) { self.session = session }
  func onThoughtChunk(text: String) {}
  func onAudioFrames(pcm: [Float], sampleRate: UInt32) {}
  func onTextChunk(text: String) {
    do {
      _ = try session.recoveryStatus()
      unexpected += 1
    } catch FfiError.Busy { busy += 1 } catch { unexpected += 1 }
  }
  func onDone(reason: FinishReason) { done.append(reason) }
}

@main
struct RecoveryProbe {
  static func main() throws {
    let args = Array(CommandLine.arguments.dropFirst())
    try require(args.count == 2, "pass fixture and cpu/metal/wgpu")
    let backend: BackendPreference
    switch args[1] {
    case "cpu": backend = .cpu
    case "metal": backend = .metal
    case "wgpu": backend = .gpu
    default: throw NSError(domain: "backend", code: 1)
    }
    let wide: UInt64 = (1 << 40) + 7
    let reasons: [KvRewindFailure] = [
      .outOfBounds(requested: wide, current: 3), .compressed, .nonCausal,
      .missingConvolutionCheckpoint(layer: 2, position: wide),
      .invalidCacheLayout(layer: 5, detail: "unequal rows"), .backendUnsupported,
      .unknown(detail: "future rewind cause"),
    ]
    for reason in reasons {
      let original = SessionRecoveryStatus(
        usable: false, position: 3,
        lastIngestRecovery: IngestRecovery(
          outcome: .unusable, rewindError: reason,
          resetError: .OutOfMemory(requestedBytes: wide)))
      let decoded = try FfiConverterTypeSessionRecoveryStatus_lift(
        FfiConverterTypeSessionRecoveryStatus_lower(original))
      try require(decoded == original, "nested diagnostic wire payload")
    }
    var cases = 0
    for compression: KvCompression in [.none, .f16, .turboQuant(seed: 7, keys: true, values: true)]
    {
      for operation in 0..<3 {
        let model = try ModelLoader(
          source: .path(path: args[0]), config: EngineConfig(backend: backend)
        ).buildGenerative()
        let referenceModel = try ModelLoader(
          source: .path(path: args[0]), config: EngineConfig(backend: backend)
        ).buildGenerative()
        let config = SessionConfig(
          maxSeqLen: 32, kvCompression: compression, seed: 1361, ubatchSize: 1)
        let actual = try model.createSession(config: config)
        let reference = try referenceModel.createSession(config: config)
        try require(try actual.recoveryStatus().lastIngestRecovery == nil, "initial diagnostic")
        try actual.appendTokens(tokens: [0, 1])
        do {
          try actual.sendMessage(message: UserMessage(text: String(repeating: "a", count: 64)))
          throw NSError(domain: "expected overflow", code: 1)
        } catch FfiError.ContextOverflow(let max, let by) {
          try require(max == 32 && by == 34, "overflow payload")
        }
        let unchanged = try actual.recoveryStatus()
        try require(
          unchanged.usable && unchanged.position == 2
            && unchanged.lastIngestRecovery?.outcome == .unchanged, "unchanged")
        try actual.appendTokens(tokens: [1])
        try require(
          try actual.recoveryStatus().lastIngestRecovery?.outcome == .unchanged,
          "raw report lifetime")
        try actual.reset()
        try require(
          try actual.recoveryStatus().lastIngestRecovery == nil, "explicit reset clears report")
        try actual.appendTokens(tokens: [0, 1])
        actual.cancel()
        let message = UserMessage(text: "baba")
        let failedSink = ProbeSink(actual)
        do {
          switch operation {
          case 0: try actual.sendMessage(message: message)
          case 1:
            _ = try actual.sendMessageAndGenerate(
              message: message, opts: GenerateOpts(maxTokens: 4))
          default:
            _ = try actual.sendMessageStreaming(
              message: message, opts: GenerateOpts(maxTokens: 4), sink: failedSink)
          }
          throw NSError(domain: "expected cancellation", code: 1)
        } catch FfiError.Cancelled {}
        let status = try actual.recoveryStatus()
        let expected: RecoveryOutcome =
          args[1] == "cpu" && compression != .turboQuant(seed: 7, keys: true, values: true)
          ? .restored : .reset
        try require(
          status.usable && status.lastIngestRecovery?.outcome == expected, "recovery outcome")
        try require(status.position == (expected == .restored ? 2 : 0), "recovery position")
        try require(status.lastIngestRecovery?.resetError == nil, "unexpected reset failure")
        if expected == .reset {
          try require(
            status.lastIngestRecovery?.rewindError
              == (args[1] == "cpu" ? .compressed : .backendUnsupported), "rewind cause")
        }
        if operation == 2 {
          try require(failedSink.done == [.cancelled], "streaming primary error")
        }
        // Observation must not clear cancellation; another multi-chunk message still fails.
        do {
          try actual.sendMessage(message: message)
          throw NSError(domain: "cancel latch cleared", code: 1)
        } catch FfiError.Cancelled {}
        actual.clearCancel()
        if expected == .reset { try actual.appendTokens(tokens: [0, 1]) }
        try reference.appendTokens(tokens: [0, 1])
        for session in [actual, reference] { try session.sendMessage(message: message) }
        try require(try actual.recoveryStatus().lastIngestRecovery == nil, "success clears report")
        try require(status.lastIngestRecovery?.outcome == expected, "owned snapshot changed")
        let options = GenerateOpts(maxTokens: 4, temperature: 1.0)
        let actualTokens = try actual.generate(opts: options).tokens
        let referenceTokens = try reference.generate(opts: options).tokens
        try require(
          actualTokens == referenceTokens && actualTokens.count == 4, "fresh continuation")
        // A same-length wrong prefix must be observable, or token parity is vacuous.
        try reference.reset()
        try reference.appendTokens(tokens: [1, 0])
        try reference.sendMessage(message: message)
        let wrongPrefixTokens = try reference.generate(opts: options).tokens
        try require(
          actualTokens != wrongPrefixTokens, "continuation oracle cannot detect a wrong prefix")
        let sink = ProbeSink(actual)
        _ = try actual.generateStreaming(opts: GenerateOpts(maxTokens: 2), sink: sink)
        try require(
          sink.busy > 0 && sink.unexpected == 0 && sink.done.count == 1, "callback reentrancy")
        cases += 1
      }
    }
    print("passed \(cases) recovery cases: \(args[1])")
  }
}
